File: src/main/java/cz/vhromada/catalog/mapper/MovieMapper.java
package cz.vhromada.catalog.mapper;

import cz.vhromada.catalog.entity.Movie;

import org.mapstruct.Mapper;

/**
 * An interface represents mapper for movie.
 *
 * @author Vladimir Hromada
 */
@Mapper
public interface MovieMapper {

    /**
     * Maps entity movie to domain movie.
     *
     * @param source entity movie
     * @return mapped domain movie
     */
    cz.vhromada.catalog.domain.Movie map(Movie source);

    /**
     * Maps domain movie to entity movie.
     *
     * @param source domain movie
     * @return mapped entity movie
     */
    Movie mapBack(cz.vhromada.catalog.domain.Movie source);

}

File: src/main/kotlin/cz/vhromada/catalog/mapper/MovieMapper.kt
package cz.vhromada.catalog.mapper

import cz.vhromada.catalog.entity.Movie
import cz.vhromada.common.mapper.Mapper
import org.springframework.stereotype.Component

/**
 * A class represents mapper for movie.
 *
 * @author Vladimir Hromada
 */
@Component("movieMapper")
class MovieMapper(
        private val mediumMapper: MediumMapper,
        private val genreMapper: GenreMapper) : Mapper<Movie, cz.vhromada.catalog.domain.Movie> {

    override fun map(source: Movie): cz.vhromada.catalog.domain.Movie {
        return cz.vhromada.catalog.domain.Movie(
                id = source.id,
                czechName = source.czechName!!,
                originalName = source.originalName!!,
                year = source.year!!,
                language = source.language!!,
                subtitles = source.subtitles!!.filterNotNull(),
                media = mediumMapper.map(source.media!!.filterNotNull()),
                csfd = source.csfd,
                imdbCode = source.imdbCode,
                wikiEn = source.wikiEn,
                wikiCz = source.wikiCz,
                picture = source.picture,
                note = source.note,
                position = source.position,
                genres = genreMapper.map(source.genres!!.filterNotNull()),
                audit = null)
    }

    override fun mapBack(source: cz.vhromada.catalog.domain.Movie): Movie {
        return Movie(
                id = source.id,
                czechName = source.czechName,
                originalName = source.originalName,
                year = source.year,
                language = source.language,
                subtitles = source.subtitles,
                media = mediumMapper.mapBack(source.media),
                csfd = source.csfd,
                imdbCode = source.imdbCode,
                wikiEn = source.wikiEn,
                wikiCz = source.wikiCz,
                picture = source.picture,
                note = source.note,
                position = source.position,
                genres = genreMapper.mapBack(source.genres))
    }

}
